### ISSUE TITLE
Non-user playlist URIs do not work in Spotify 0.9.17.8

### ISSUE THREAD

**Issue by atatdotdot** [olejon/spotcommander#134]
I was having some trouble with stability in spotcommander when using the latest client and Spotify Free, so I followed your guide and downgraded to your packaged version of client 0.9.17.8. Since then stability has been much better, but I am unable to play playlists from spotcommander.

I have tracked down what I believe to be the source of this problem; it appears that the URIs used internally when searching for and browsing playlists are of the form spotify:playlist:xxxxxx, but that the 0.9.17.8 client only works with spotify:user:xxxxxx:playlist:xxxxxx.

If I run:

```qdbus org.mpris.MediaPlayer2.spotify /org/mpris/MediaPlayer2 org.mpris.MediaPlayer2.Player.OpenUri spotify:playlist:6XbZD7guSDJ6QC3edZRKlM```

the client flashes a message "Link not found", but if I run:

```qdbus org.mpris.MediaPlayer2.spotify /org/mpris/MediaPlayer2 org.mpris.MediaPlayer2.Player.OpenUri spotify:user:greenbeltfestival:playlist:6XbZD7guSDJ6QC3edZRKlM```

it works.

Perhaps it would be possible to just always use the spotify:user:xxxxxx:playlist:xxxxxx style internally? I presume all playlists have a user who made them, and getting that isn't hard, and that the new client accepts the old format?

I'm happy to do a patch but didn't want to start if I'm barking up the wrong tree.

**Comment by atatdotdot** [comment]
I have realized that this appears to be caused by the same underlying issue as https://github.com/olejon/spotcommander/issues/132; sorry for the duplication.

**Comment by atatdotdot** [comment]
This patch fixes the problem for playlists that come back from searches, but not stored playlists. I wonder if there is a better way to go about this?

```--- functions.php.old	2019-06-19 12:04:59.705332692 +0000
+++ functions.php	2019-06-19 12:06:57.722039988 +0000
@@ -1342,7 +1342,7 @@
 
 				$return['playlists'][$i]['name'] = $playlist['name'];
 				$return['playlists'][$i]['user'] = $playlist['owner']['id'];
-				$return['playlists'][$i]['uri'] = $playlist['uri'];
+				$return['playlists'][$i]['uri'] = 'spotify:user:' . $playlist['owner']['id'] . ':playlist:' . $playlist['id'];
 				$return['playlists'][$i]['cover_art'] = $playlist['images'][0]['url'];
 
 				$i++;
```

**Comment by olejon** [comment]
Yes, it seems like Spotify has changed the Playlist URIs `spotify:` and URLs `open.spotify.com` and `play.spotify.com`.

Best way to deal with this is to fix the functions that uses RegEx to determine if URI/URL is a track, album, artist, playlist or user in both PHP and JS.

* Look here for PHP: https://github.com/olejon/spotcommander/blob/master/spotcommander/functions.php#L2118
* Look here for JS: https://github.com/olejon/spotcommander/blob/master/spotcommander/js/functions.js#L3418

You need to change (PHP uses _underscores_ and JS camelCase):
* get uri type
* uri to url
* url to uri

Browse I need to fix as it goes through my API. I use the same RegEx there.

**Comment by olejon** [comment]
Spotify has changed all THEIR, or maybe all found in the API, playlists to another format: `spotify:playlist:37i9dQZF1DWUJF24WXSSyO

Try to play these two playlists with D-Bus, without modifying:

* `spotify:user:indiefolkradio:playlist:7r1SZzTwzqtqgDYvRbqGhz`
* `spotify:playlist:37i9dQZF1DWUJF24WXSSyO`

**If the last does not work, one must probably add the user `spotify` to it, like so:**

* `spotify:user:spotify:playlist:37i9dQZF1DWUJF24WXSSyO`

Does that work?

**If so, there is a GLOBAL place you can modify this, in `main.php`:**

* See: https://github.com/olejon/spotcommander/blob/master/spotcommander/main.php#L194
* Edit `main.php` on your PC
* Make the `$data` variable is edited to convert ALL *without* a username to contain `spotify:user:spotify:playlist:[22_CHARACTERS]`
* Basically adding in after `spotify:` the text  `user:spotify`

Tell me if you don't know how to do this.

**For proper detection of playlists, I fixed RegEx on the server the one place it is used:**
```
if(preg_match('/^spotify:user:[^:]+:playlist:\w{22}$/', $uri) || preg_match('/^spotify:playlist:\w{22}$/', $uri) || <LEAVE_THIS_ALONE_FOR_NOW>
```

Should be enough for you to understand how to add it to the other functions that uses it to detect "if playlist" in **both** the PHP and JS files I linked to, namely `functions.php` and `js/functions.js` on your PC.

**Please do this when connected to SpotCommander on your PC:**
* Please go to Settings (Side menu) > Clear cache (Almost at the bottom)
* Press CHECK icon in the upper right corner to refresh
* I think now Cover Art should start appearing for your Playlists etc, is that right?

**Comment by olejon** [comment]
EDITED the comment above. Please read it here, not what comes to your e-mail.
